### PR TITLE
Implement Nullish Coalescing operator

### PR DESCRIPTION
**Pull request by rhuanjl**
This PR takes work begun by @boingoing in #6254 and completes it to implement the stage 4 tc39 proposal for a nullish coalescing operator.

Proposal repository: https://github.com/tc39/proposal-nullish-coalescing
The implementations passes most relevant test262 test cases - exception being ones also involving tail call optimisations which is a separate JS feature that CC does not support.

A switch `ESNullishCoalescingOperator` is included but set to true by default.

The basic logic of this implementation is fairly simple - though one key complexity is that `a ?? b || c` is not permitted without brackets around one of the two operations, this required some extra tracking in the parser.

Within the bytecode emitter the noalish coalescer is de-sugered so that:
`result = node1 ?? node2;`
becomes
```
result = node1;
if (result == null) {
  result = node2;
}
```

@ppenzin @Fly-Style please could one of you take a look?

### CODE REVIEW

File: lib/Parser/Parse.cpp [hunk 1]
```diff
@@ -279,6 +279,7 @@ LPCWSTR Parser::GetTokenString(tokens token)
     case tkColon: return _u(":");
     case tkLogOr: return _u("||");
     case tkLogAnd: return _u("&&");
+    case tkCoalesce: return _u("??");
```
Comment (rhuanjl) [inline lib/Parser/Parse.cpp:282]: This switch converts parser tokens back into strings for error messages.

File: lib/Parser/Parse.cpp [hunk 2]
```diff
@@ -3237,7 +3238,8 @@ ParseNodePtr Parser::ParseTerm(BOOL fAllowCall,
     _Out_opt_ BOOL* pfCanAssign /*= nullptr*/,
     _Inout_opt_ BOOL* pfLikelyPattern /*= nullptr*/,
     _Out_opt_ bool* pfIsDotOrIndex /*= nullptr*/,
-    _Inout_opt_ charcount_t *plastRParen /*= nullptr*/)
+    _Inout_opt_ charcount_t *plastRParen /*= nullptr*/,
+    _Out_opt_ bool* looseCoalesce /*= nullptr*/)
```
Comment (rhuanjl) [inline lib/Parser/Parse.cpp:3242]: The looseCoalesce out parameter enables tracking of unbracketed coalescing operators for the syntax error when they are combined with ?? or ||.

File: lib/Parser/Parse.cpp [hunk 3]
```diff
@@ -9299,8 +9313,26 @@ ParseNodePtr Parser::ParseExpr(int oplMin,
...
             IdentToken token;
+            bool coalescing = false;
+
             ParseNode* pnode2 = ParseExpr<buildAST>(
-                opl, nullptr, fAllowIn, FALSE, pNameHint, &hintLength, &hintOffset, &token, false, nullptr, plastRParen);
+                opl, nullptr, fAllowIn, FALSE, pNameHint, &hintLength, &hintOffset, &token, false, nullptr, plastRParen, &coalescing);
+
+            if (nop == knopLogAnd || nop == knopLogOr)
+            {
+                if (localCoalesce || coalescing)
```
Comment (rhuanjl) [inline lib/Parser/Parse.cpp:9323]: Here's the check for the syntax error - `localCoalesce` will be true if the left hand side of this operation is an unbracketed ??,  `coalescing` will be true if the right hand side is.

Aware the variable names aren't the best - open to alternatives.

File: lib/Parser/Parse.cpp [hunk 4]
```diff
@@ -9299,8 +9313,26 @@ ParseNodePtr Parser::ParseExpr(int oplMin,
...
+                {
+                    Error(ERRCoalesce);
+                }
+            }
+            else if (nop == knopCoalesce)
+            {
+                localCoalesce = true;
+                if (looseCoalesce != nullptr)
+                {
+                    *looseCoalesce = true;
```
Comment (rhuanjl) [inline lib/Parser/Parse.cpp:9333]: Here we mark that the current operation is a coalescing operation to feed up into the above check at a different level.

File: lib/Parser/Parse.cpp [hunk 5]
```diff
@@ -3440,6 +3442,11 @@ ParseNodePtr Parser::ParseTerm(BOOL fAllowCall,
 
         ChkCurTok(tkRParen, ERRnoRparen);
 
+        if (looseCoalesce != nullptr)
+        {
+            *looseCoalesce = false;
```
Comment (rhuanjl) [inline lib/Parser/Parse.cpp:3447]: The last passed statement was wrapped in brackets - therefore it is not a "loose" coalesce.

File: lib/Parser/perrors.h [hunk 6]
```diff
@@ -6,85 +6,85 @@
...
-LSC_ERROR_MSG( 1075, ERRInvalidAssignmentTarget, "Invalid destructuring assignment target")
-LSC_ERROR_MSG( 1076, ERRFormalSame, "Duplicate formal parameter names not allowed in this context")
-LSC_ERROR_MSG( 1077, ERRDestructNotInit, "Destructuring declarations cannot have an initializer")
-// 1078 -- removed
+LSC_ERROR_MSG(1073, ERRYieldInTryCatchOrFinally, "'yield' expressions are not allowed in 'try', 'catch', or 'finally' blocks")
+LSC_ERROR_MSG(1074, ERRConstructorCannotBeGenerator, "Class constructor may not be a generator")
+LSC_ERROR_MSG(1075, ERRInvalidAssignmentTarget, "Invalid destructuring assignment target")
+LSC_ERROR_MSG(1076, ERRFormalSame, "Duplicate formal parameter names not allowed in this context")
+LSC_ERROR_MSG(1077, ERRDestructNotInit, "Destructuring declarations cannot have an initializer")
+LSC_ERROR_MSG(1078, ERRCoalesce, "Coalescing operator '\?\?' not permitted unparenthesized inside '||' or '&&' expressions")
```
Comment (rhuanjl) [inline lib/Parser/perrors.h:87]: New error message for the new syntax error. The changes above in this file are just white space.

File: lib/Runtime/ByteCode/ByteCodeEmitter.cpp [hunk 7]
```diff
@@ -11724,6 +11724,32 @@ void Emit(ParseNode* pnode, ByteCodeGenerator* byteCodeGenerator, FuncInfo* func
         ENDSTATEMENET_IFTOPLEVEL(isTopLevel, pnode);
         break;
     }
+    // The Coalescing operator resolves to the left hand side if it is not null or undefined
+    // In that case the right hand side is not evaluated
+    // If the left hand side is null or undefined it resolves to the right hand side
+    // PTNODE(knopCoalesce     , "??"        ,None    ,Bin  ,fnopBin)
+    case knopCoalesce:
```
Comment (rhuanjl) [inline lib/Runtime/ByteCode/ByteCodeEmitter.cpp:11731]: This case in the bytecode emitter is the run time implementation - the command is fully de-sugared here, so no new opcodes were needed.

**Comment by fatcerberus** [comment]
Word of warning that per the DOM spec, `document.all` is falsy but is expected to be treated as truthy by operators like `??` - Edge doesn’t use Chakra anymore so this is less important but something to keep in mind since I don’t think embedders can emulate this behavior in userland.

**Comment by rhuanjl** [comment]
> Word of warning that per the DOM spec, `document.all` is falsy but is expected to be treated as truthy by operators like `??` - Edge doesn’t use Chakra anymore so this is less important but something to keep in mind since I don’t think embedders can emulate this behavior in userland.

Falsely shouldn't be a problem - only if tests with `== null` if it does that's bizarre BUT I've implemented what the spec says and this isn't currently used in a browser anyway so I think we just roll with it.

**Comment by fatcerberus** [comment]
Yeah, oddly enough `document.all == null` is actually true even though it’s an object, verified in Chrome console. Apparently it’s some compatibility baggage from ancient Internet Explorer.

I don’t think it’s a major issue in practice, like you said. Just wanted to state it for the record since embedders have no way to emulate this behavior.

**Review by ppenzin — APPROVED** [review]

**Comment by ppenzin** [comment]
Doing the usual clopenning to clear persistent CI failure.

**Comment by rhuanjl** [comment]
> Yeah, oddly enough `document.all == null` is actually true even though it’s an object, verified in Chrome console. Apparently it’s some compatibility baggage from ancient Internet Explorer.
> 
> I don’t think it’s a major issue in practice, like you said. Just wanted to state it for the record since embedders have no way to emulate this behavior.

Thinking about it - if any browser does embed this and if a developer wants to do `document.all ?? something` and is confused at getting `something` as the result that's on them - the sort of special casing required to avoid that is IMO not worth the performance cost it would entail for all other uses.

**Comment by fatcerberus** [comment]
It turns out that the special case under discussion is actually formally encoded in the ES spec, see https://tc39.es/ecma262/#sec-IsHTMLDDA-internal-slot

**Comment by rhuanjl** [comment]
> It turns out that the special case under discussion is actually formally encoded in the ES spec, see https://tc39.es/ecma262/#sec-IsHTMLDDA-internal-slot

Yeah I read it - and I don't think it's worth supporting for now, to handle it in this operation could either:
a) implement a whole new opcode OR
b) use two strict equal comparisons instead of one normal comparison

(a) would be a lot of extra code and (b) would be a performance hit - if anyone actually makes a new webbrowser with CC and wants spec perfect support for window.all we can revisit.

**Comment by fatcerberus** [comment]
Alright, that makes sense. Just wanted to make sure everything was on the table first before a decision was made.  I’ll duck out now 🦆

**Comment by ppenzin** [comment]
Closing and reopening in hopes to clear the CI failure.

CC @divmain ;)

**Comment by rhuanjl** [comment]
@ppenzin interestingly there's an option to override the CI and merge - I'm guessing there's no way to retrigger it now the repository has been transferred - considering it passed everything other than windows 7 - think we should just merge it?

**Comment by ppenzin** [comment]
@rhuanjl agree - we should merge it.

**Pull request merged**